sql/core/src/main/scala/org/apache/spark/sql/execution/dynamicpruning/PartitionPruning.scala

```diff
@@ -167,44 +167,10 @@ object PartitionPruning extends Rule[LogicalPlan] with PredicateHelper with Join
     }
 
     val estimatePruningSideSize = filterRatio * partPlan.stats.sizeInBytes.toFloat
-    val overhead = calculatePlanOverhead(otherPlan)
+    val overhead = collectLeafPlanOverhead(otherPlan).sum.toFloat
     estimatePruningSideSize > overhead
   }
 
-  /**
-   * Calculates a heuristic overhead of a logical plan. Normally it returns the total
-   * size in bytes of all scan relations. We don't count in-memory relation which uses
-   * only memory.
-   */
-  private def calculatePlanOverhead(plan: LogicalPlan): Float = {
-    val (cached, notCached) = plan.collectLeaves().partition(p => p match {
-      case _: InMemoryRelation => true
-      case _ => false
-    })
-    val scanOverhead = notCached.map(_.stats.sizeInBytes).sum.toFloat
-    val cachedOverhead = cached.map {
-      case m: InMemoryRelation if m.cacheBuilder.storageLevel.useDisk &&
```

Comment:
the cached relation should be materialized right ? otherwise its statistics should be same with normal plan.

Reply:
this method moved into the new trait.